### PR TITLE
In local invocations, do not require -PowerShellVersion 5.1 for Get-FormatData in order to see all format data.

### LINKED ISSUE #4237: Get-FormatData doesn't report formatting data for some types for which formatting data ships with PowerShell.
I haven't looked into the logic behind which types' formatting data isn't being reported, but two examples are below.

A counter example is `Get-FormatData System.Management.Automation.AliasInfo`, which works as expected.

Steps to reproduce
------------------

```powershell
Get-FormatData System.IO.DirectoryInfo, System.IO.FileInfo
```

Expected behavior
-----------------

`[System.Management.Automation.ExtendedTypeDefinition]` instances describing the formatting, as usual.

Actual behavior
---------------

No output is produced; `$?` indicates `$True`.

Yet, formatting data is clearly in effect.

In _Windows PowerShell_, it can be discovered as follows:

```powershell
Select-String -List System.IO.DirectoryInfo, System.IO.FileInfo $PSHOME/*.ps1xml 
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.3 on macOS 10.12.5
PowerShell Core v6.0.0-beta.3 on Ubuntu 16.04.1 LTS
PowerShell Core v6.0.0-beta.3 on Microsoft Windows 10 Pro (64-bit; v10.0.14393)
Windows PowerShell v5.1.15063.413 on Microsoft Windows 10 Pro (64-bit; v10.0.15063)
```

### PR DESCRIPTION
**Pull request by mklement0**
# PR Summary

Fix #4237

## PR Context

In order to support older remoting clients, certain format data that requires v5.1+ is not returned from `Get-FormatData` by default (only if `-PowerShellVersion 5.1` is passed).

In _local_ invocations, this behavior is unexpected: all available format data should be returned.

This PR fixes that, both for direct local invocations and for invocations in background jobs.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---subfmission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5076
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 [hunk 1]
```diff
@@ -3,22 +3,36 @@
...
-    It "Can get format data requiring '-PowerShellVersion 5.1'" {
-        $format = Get-FormatData System.IO.FileInfo -PowerShellVersion 5.1
-        $format.TypeNames | Should -HaveCount 2
-        $format.TypeNames[0] | Should -BeExactly "System.IO.DirectoryInfo"
-        $format.TypeNames[1] | Should -BeExactly "System.IO.FileInfo"
-        $format.FormatViewDefinition | Should -HaveCount ($IsWindows ? 4 : 5)
+    Context "Local use: Can get format data requiring v5.1 by default" {
+        $cmds =
+            @{ cmd = { Get-FormatData System.IO.FileInfo } },
+            @{ cmd = { (Get-FormatData System.IO.FileInfo &) | Receive-Job -Wait -AutoRemoveJob } }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: Please put in `BeforeAll` block.

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: Is it reliable to use `System.IO.FileInfo` format if it could be changed?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 [hunk 2]
```diff
@@ -3,22 +3,36 @@
...
+        It "Can get format data requiring v5.1 with <cmd>" -TestCases $cmds {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]:
```suggestion
        It "Can get format data requiring v5.1 with '<cmd>'" -TestCases $cmds {
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 [hunk 3]
```diff
@@ -3,22 +3,36 @@
...
+            param([scriptblock] $cmd)
+            $format = & $cmd
+            $format.TypeNames | Should -HaveCount 2
+            $format.TypeNames[0] | Should -BeExactly "System.IO.DirectoryInfo"
+            $format.TypeNames[1] | Should -BeExactly "System.IO.FileInfo"
+            $format.FormatViewDefinition | Should -HaveCount ($IsWindows ? 4 : 5)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: The same question about reliability.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 (hunk 1 above)
Comment (mklement0) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: Re `BeforeAll`: 👍

Re reliability: I agree - the point here is not to test the specifics of the format data returned (but that's how the test was originally written).

I've replaced with a simple `(& $cmd).TypeNames | Should -Contain 'System.IO.FileInfo'` - which hopefully everyone is happy with.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 (hunk 2 above)
Comment (mklement0) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: Surprisingly, Pester actually stringifies script blocks _with_ the enclosing `{` and `}`, so you'll get something like:

```none
When remoting, { Get-FormatData System.IO.FileInfo } should return $true for a null-output test
```

So I think that's sufficient - no need to quote as well.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 (hunk 3 above)
Comment (mklement0) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: See comment above.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 (hunk 1 above)
Comment (mklement0) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: As for whether to _fundamentally_ use `System.IO.FileInfo` in the tests: That type is known as one that mustn't be returned to older clients, so I think it is a reliable choice.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs [hunk 4]
```diff
@@ -97,9 +97,15 @@ protected override void BeginProcessing()
...
-                               (PowerShellVersion.Major == 5 && PowerShellVersion.Minor < 1);
+            // During remoting, remain compatible with older clients by default.
+            // Remoting detection: Automatic variable $PSSenderInfo is defined in true remoting contexts and also in background jobs.
+            //                     However, its Boolean .PSShowComputerName NoteProperty is true only in remoting contexts.
+            bool remoting = (GetVariableValue("PSSenderInfo") as PSObject)?.Properties["PSShowComputerName"]?.Value is bool value && value;
+            bool writeOldWay = remoting && (
+                                PowerShellVersion == null ||
+                                PowerShellVersion.Major < 5 ||
+                                (PowerShellVersion.Major == 5 && PowerShellVersion.Minor < 1)
+                                );
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: Arbitrary indents tend to be a bit unwieldy. Do you think that perhaps something a bit more along these lines might be more readable?

```csharp
bool writeOldWay = remoting
    && (PowerShellVersion == null 
        || PowerShellVersion.Major < 5
        || (PowerShellVersion.Major == 5 && PowerShellVersion.Minor < 1));
```

Perhaps also we could rename `remoting` to something like `isRemoting` or `inRemoteSession` for additional clarity?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 [hunk 5]
```diff
@@ -3,22 +3,36 @@
...
-        $format = Get-FormatData System.IO.FileInfo -PowerShellVersion 5.1
-        $format.TypeNames | Should -HaveCount 2
-        $format.TypeNames[0] | Should -BeExactly "System.IO.DirectoryInfo"
-        $format.TypeNames[1] | Should -BeExactly "System.IO.FileInfo"
-        $format.FormatViewDefinition | Should -HaveCount ($IsWindows ? 4 : 5)
+    Context "Local use: Can get format data requiring v5.1 by default" {
+        BeforeAll {
+            $cmds =
+                @{ cmd = { Get-FormatData System.IO.FileInfo } },
+                @{ cmd = { (Get-FormatData System.IO.FileInfo &) | Receive-Job -Wait -AutoRemoveJob } }
```
Comment (vexx32) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: This might prove a little fragile for future maintenance if we end up wanting to add further test cases. Can we wrap this array with `@()` and drop the commas so it's a bit less hassle to add and remove entries without breaking the syntax? 🙂 

Something like this:
```powershell
$cmds = @(
    @{ cmd = { <# ... #> } }
    @{ cmd = { <# ... #> } }
)
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1 [hunk 6]
```diff
@@ -3,22 +3,36 @@
...
 
-    It "Should return nothing for format data requiring '-PowerShellVersion 5.1' and not provided" {
-        Get-FormatData System.IO.FileInfo | Should -BeNullOrEmpty
+    Context "Remote use: Don't get format data requiring v5.1 by default" {
+        BeforeAll {
+            # Simulate a remoting session.
+            $PSSenderInfo = [pscustomobject] @{ PSShowComputerName = $true }
+            $cmds =
+                @{ cmd = { Get-FormatData System.IO.FileInfo }; shouldBeNull = $true },
+                @{ cmd = { Get-FormatData System.IO.FileInfo -PowerShellVersion 5.1 }; shouldBeNull = $false }
```
Comment (vexx32) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Get-FormatData.Tests.ps1]: Same comment here as above. 🙂

**Review by vexx32 — COMMENTED** [review]
Just some minor style comments. Looking food so far, thanks for PRing this one @mklement0! 💖 😊

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: I agree that the formatting should be fixed but the renaming will be reported as CodeFactor issue ""using Bulgarian prefixes".

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: All good points, @vexx32 - fixed.

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: @iSazonov: I went with `inRemoteSession` - will that be a problem? I've heard of _Hungarian_ prefixes - what are _Bulgarian_ ones?

Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: Didn't we at some point add `in` or `is` to the permitted Hungarian prefixes for codefactor? I may be misremembering. Worst case we can use `calledFromRemoteSession` but that is a bit wordy. Other suggestions also welcome, of course! 😊

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: It looks like CodeFactor didn't complain about `inRemoteSession`.

**Comment by SteveL-MSFT** [comment]
@mklement0 can you look at the test failures?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs [hunk 7]
```diff
@@ -97,9 +97,14 @@ protected override void BeginProcessing()
         /// </summary>
         protected override void ProcessRecord()
         {
-            bool writeOldWay = PowerShellVersion == null ||
-                               PowerShellVersion.Major < 5 ||
-                               (PowerShellVersion.Major == 5 && PowerShellVersion.Minor < 1);
+            // During remoting, remain compatible with older clients by default.
+            // Remoting detection: Automatic variable $PSSenderInfo is defined in true remoting contexts and also in background jobs.
+            //                     However, its Boolean .PSShowComputerName NoteProperty is true only in remoting contexts.
+            bool inRemoteSession = (GetVariableValue("PSSenderInfo") as PSObject)?.Properties["PSShowComputerName"]?.Value is bool value && value;
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: This doesn't look right to me.  The `PSSenderInfo` variable in the session is always of `PSSenderInfo` type.  It is only a `PSObject` type if it goes through the serialization system.  So retrieving the built in variable here as `PSObject` type will always be null and `inRemoteSession` will always be false.

The only way I can think of distinguishing between remote session and background job (remote) session is to change the `PSSenderInfo.ConnectionString` to something different for background jobs (since it is meaningless for background jobs), maybe just `PSSenderInfo.ConnectionString = "BackgroundJob"`.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 7 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: That sounds like a fairly effective solution. 🤔

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: Because it is masked in Settings.StyleCop but it is not best practice.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 7 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: We have a sample:
https://github.com/PowerShell/PowerShell/blob/ee4eba5348739b6ef17cf2069685cc2d1815a62e/src/System.Management.Automation/engine/GetCommandCommand.cs#L492-L504

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: You're right, @PaulHigin. Apart from the `PSObject` problem, looking for this property on the _server_ side is the wrong approach anyway.

Setting `.ConnectionString` to something different makes perfect sense - I would just set it to the empty string, given that there's no connection involved.

Currently, it it is `https://localhost` - any concerns about changing it?

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: @iSazonov: The linked approach doesn't distinguish between true remoting and background jobs, which we need to make.

**Comment by mklement0** [comment]
@SteveL-MSFT, I'll focus on fixing the fundamental problem first. A preliminary investigation showed that the problem was just one of _duplicates_ in the output, strangely; applying `Select-Object -Unique` before counting / comparing made the problem go away.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 7 above)
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: @mklement0 I still believe the check should be made on the server side.  If I understand the problem correctly, we need to use the 'old format way' when talking remotely to older clients.  So I think we can make the right decision on the server by looking at the PSSenderInfo.  Something like:

```
bool writeOldWay = false;
var psSenderInfo = GetVariableValue("PSSenderInfo");
if (psSenderInfo != null)
{
    var clientPSVersion = psSenderInfo.ApplicationArguments?.PSVersionTable?.PSVersion
    writeOldWay = (clientPSVersion != null) && (clientVersion.Major < 5 || (clientVersion.Major == 5 && clientVersion.Minor < 1));
}
...
```

File: src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs [hunk 8]
```diff
@@ -154,7 +154,8 @@ public PSPrimitiveDictionary ApplicationArguments
         {
             get { return _applicationArguments; }
 
-            internal set { _applicationArguments = value; }
+            // Conceptually `internal`, but public to support simulating remoting sessions during testing.
+            set { _applicationArguments = value; }
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/remoting/fanin/PSPrincipal.cs]: Please use a test hook. See `InternalTestHooks`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs (hunk 7 above)
Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: @PaulHigin: Absolutely, the test needs to be on the server side (the problem with my previous attempt was to approach the problem from the client side).

Thanks for pointing me to the fact that `$psSenderInfo.ApplicationArguments` actually contains the client `$PSVersionTable`.

I've rewritten the test as follows:

* If `-PowerShellVersion` is used to explicitly indicate a client version, use it compare it to `5.1` - irrespective of whether the invocation is local or not.

* Otherwise, if `$PSSenderInfo` is present, examine `$psSenderInfo.ApplicationArguments.PSVersionTable.PSVersion` - i.e., the remoting client's version - and test that against `5.1`. Treat the case where `$PSSenderInfo` is present, but the client version cannot be extracted as the pre-`5.1` case, to be safe (not sure if that could ever happen).

* Otherwise (purely local invocation without `-PowerShellVersion`), return _all_ formats.

It's fine to then not to have to distinguish between true remoting and background jobs.

The above leads me  to think that introducing `-PowerShellVersion` may never have been needed, because the client version information could always have been derived from `$psSenderInfo.ApplicationArguments.PSVersionTable.PSVersion` (or its absence), correct?

---

Separately - and this will have to become separate issues:

I still think it would make sense to have `.ConnectionString` reflect whether code is executing in a remote session or in a background job.

* You can _almost_ infer it already, namely by testing for `http://localhost`, whereas WinRM-based invocations contain a port number and client version number; e.g., `http://localhost:5985/wsman?PSVersion=5.1.18362.145

* However, `http://localhost` is _also_ the value in _SSH_-based remoting sessions - which is a problem in itself. If this gets fixed, and there's a concern that `.ConnectionString` should otherwise not be changed for the sake of backward compatibility, `http://localhost` could be used (though it would be unexpected, given that there's no HTTP transport involved, I presume).

Comment (mklement0) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/FormatAndOutput/common/GetFormatDataCommand.cs]: Please see #11292 and #11293.

**Comment by mklement0** [comment]
@SteveL-MSFT, looking into this further my conclusion is that this PR has simply unmasked _an unrelated problem that should be investigated separately_.

For now, I've restored the semantics of the original test by passing `-PowerShellVersion 5.0` to the `Get-FormatData` commands in the original tests, which makes them succeed.

As for what may be going on:

The current format-data definitions contain _duplicates_:

Some types are covered _both_ by a dedicated definition (where `.TypeNames` references a _single_ type) _and_ by a _multi-type_ definition (where `.TypeNames` contains multiple type names).

```powershell
$dataWithMultipleTypes = Get-FormatData -PowerShellVersion 5.1 | Where { $_.TypeNames.Count -gt 1 }
$dataWithSingleType = Get-FormatData -PowerShellVersion 5.1 | Where { $_.TypeNames.Count -eq 1 }

# Find all single-type definitions also covered by multiple-type definitions.
$dataWithSingleType | ? { foreach ($fd in $dataWithMultipleTypes) { if ($fd.TypeNames -contains $_) { return $true } } }
```

This currently yields:

```none
TypeNames                                                        FormatViewDefinition
---------                                                        --------------------
{System.Security.Cryptography.X509Certificates.X509Certificate2} {ThumbprintTable}
{System.Management.Automation.ApplicationInfo}                   {ApplicationInfo, System.Management.Automation.ApplicationInfo}
{System.Management.Automation.ScriptInfo}                        {ScriptInfo, System.Management.Automation.ScriptInfo}
{System.Management.Automation.ExternalScriptInfo}                {ExternalScriptInfo}
{System.Management.Automation.FunctionInfo}                      {FunctionInfo}
{System.Management.Automation.FilterInfo}                        {FilterInfo}
{System.Management.Automation.AliasInfo}                         {AliasInfo, System.Management.Automation.AliasInfo}
{System.Management.Automation.CmdletInfo}                        {System.Management.Automation.CmdletInfo}
{System.Management.Automation.PSDriveInfo}                       {System.Management.Automation.PSDriveInfo}
```

* Is there a good reason for this duplication?
* If so, what are the precedence rules, and is `Update-FormatData` expected to load the duplicates too?

File: src/System.Management.Automation/engine/Utils.cs [hunk 9]
```diff
@@ -2090,6 +2091,25 @@ public static bool TestImplicitRemotingBatching(string commandPipeline, System.M
...
+        /// <summary>
+        /// Constructs a custom PSSenderInfo instance that can be assigned to $PSSenderInfo
+        /// in order to simulate a remoting session with respect to the $PSSenderInfo.ConnectionString (connection URL)
+        /// and $PSSenderInfo.ApplicationArguments.PSVersionTable.PSVersion (the remoting client's PowerShell version).
+        /// See Get-FormatDataTest.ps1.
+        /// </summary>
+        /// <param name="url">The connection URL to reflect in the returned instance's ConnectionString property.</param>
+        /// <param name="clientVersion">The version number to report as the remoting client's PowerShell version.</param>
+        /// <returns>The newly constructed custom PSSenderInfo instance.</returns>
+        public static PSSenderInfo GetCustomPSSenderInfo(string url, Version clientVersion)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Utils.cs]: /cc @PaulHigin @TravisEz13  Please review the public API.

**Comment by SteveL-MSFT** [comment]
@mklement0 adding `| % { Get-FormatData -PowerShellVersion 5.1 -TypeName $_.TypeNames }` to your example, it seems that the duplication is for a base type and derived types.  So that seems to be fine.  As for ordering, with the exception of Out-of-Band types (like errors), it should go in order of the psobject.typenames list.

**Review by PaulHigin — APPROVED** [review]
LGTM

**Comment by mklement0** [comment]
@SteveL-MSFT 

There is genuine duplication, with _dormant_ duplicates among the built-in formats - see #11307

I suspect that this duplication ultimately causes the round-trip `Get-FormatData` test to fail with `-PowerShellVersion 5.1` (or this fix in place), but that's a separate issue - see #11308

**Comment by mklement0** [comment]
@SteveL-MSFT, are there concerns around the public method added to `InternalTestHooks`? Apart from that, I think this is ready to be merged.

**Review by SteveL-MSFT — APPROVED** [review]
The test hooks have to be public so that the tests can call them.  They are in the `internal` namespace so users should not use them.  Since they run in the context of the user, they are effectively no different than if the user did reflection anyways.

**Comment by adityapatwardhan** [comment]
@mklement0 Please have a look at the merge conflicts

**Comment by mklement0** [comment]
@adityapatwardhan, I've resolved the merge conflict.

The PowerShell-CI-static-analysis  failure seems like a transient one ("SocketException: Network is unreachable") - not sure how to restart the test.

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry static

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-static-analysis`

**Comment by adityapatwardhan** [comment]
@mklement0 There seems to be another merge conflict. Please have a look.

**Comment by mklement0** [comment]
@adityapatwardhan , I've resolved the (trivial) conflict in favor of preferring a whitespace after the unary form of the `,` operator (`, <operand>` rather than `,<operand>`), which looks cleaner to me (it's also what auto-formatting in VSCode does).

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@mklement0 Thanks for your contribution!

**Comment by JoeSalmeri** [comment]
@SteveL-MSFT 

Please see the update I posted yesterday that still exists in PS 7.2.0 to an issue I previously reported about issues with Get-FormatData / Export-FormatData / Update-FormatData.

Here's the link with a detailed example that is 100% reproducible with PS 7.2.0

https://github.com/PowerShell/PowerShell/issues/9300

A quick summary would be that:

A roundtrip of format data using Get-FormatData / Export-FormatData / Update-FormatData results in broken output.

The 2 main issues are:

When format data has a custom control it is not returned by Get-FormatData and therefore is lost on the roundtrip.

When format data contains more than 1 TypeName, Get-FormatData seems to get both but ExportFormat data only exports the first 1 so additional types are lost in the roundtrip.

This can be seen in the PS source code

PowerShell-7.2.0/src/System.Management.Automation/FormatAndOutput/DefaultFormatters/FileSystem_format_ps1xml.cs

where the custom control is created and shared between views and where a view has more than 1 TypeName.

It is frustrating to see these problems still exist in the latest PS when they have existed and reported so long ago.

Get-FormatData / Export-FormatData / Update-FormatData are GREAT feaures but the *.format.ps1xml files should never have been removed until the replace actually was fully functional.

Fixing Export-FormatData to include all TypeNames seems like it should be a fairly easy fix since Get-FormatData appears to have them.

Fixing the CustomControl situation seems more complicated since it is shared.

Here's a thought:

DirectoryInfo / FileInfo objects have the PSParentPath note property which looks like this:

    string PSParentPath=Microsoft.PowerShell.Core\FileSystem::/home/joe

Add a new note property ParentPath

    string ParentPath=/home/joe

Then all of the places that are using the shared custom control for their GroupBy could just be changed to reference ParentPath and there would no longer be a need for the custom control!

            <GroupBy>
                <PropertyName>ParentPath</PropertyName>
            </GroupBy>

This seems like a fairly easy change and by eliminating the shared custom control you remove the more difficult situation to fix.

What do you think?

Joe
